[Web] 카카오 로그인 구현

hihibara opened the pull request:
변경사항 정리
================

1. ip 상대 주소로 변경 
 ``` java script
const WEB_IP = new URL(document.location.href);
const pathname = WEB_IP.pathname; // url-path 부분
const CALLBACK_IP =
  WEB_IP.origin + // scheme와 hosts 부분
  pathname.substring(0, pathname.lastIndexOf("/")) +
  "/CallBack.html";
```

2. <script>를 .js  파일로 이동

3. 불필요한 주석 제거

hihibara committed:
first commit

hihibara committed:
Code for login fucntion by using OUATH framework.

- this is test for second commit.

Commit:
revert branch (reset)

Commit:
기본 폴더 구조 구성

hihibara committed:
login_Ouath 네이버&카카오

hihibara committed:
docs & rename (ip 상대주소 변경, 파일 위치와 Login.html->index.html, 주석제거 )

Enoch-Kim commented on apps/web-login/js/CallBack.js:
```diff
@@ -0,0 +1,19 @@
+const callBack_URL = new URL(document.location.href);
+const index_URL = callBack_URL.origin;
+var naver_id_login = new naver_id_login(naverOuath_ID, index_URL);
+alert(index_URL);
```
디버깅할때 사용한 코드는 지워주세요 :)

Enoch-Kim commented on apps/web-login/index.html:
```diff
@@ -0,0 +1,57 @@
+          <input type="password" name="" required="" />
+          <label>Password</label>
+        </div>
+        <a href="#"> Submit </a>
```
Submit 버튼을 포함하여 UI 상의 위치들을 깔끔하게 정리해주실 수 있나요?

hihibara replied:
네 알겠습니다

Enoch-Kim commented on apps/web-login/index.html:
```diff
@@ -0,0 +1,57 @@
+        <a href="#"> Submit </a>
+      </form>
+      <div class="oauth-login">
+        <ul>
+          <li id="kakao-login-btn" onclick="kakaoLogin();">
+            <img
+              src="https://k.kakaocdn.net/14/dn/btroDszwNrM/I6efHub1SN5KCJqLm1Ovx1/o.jpg"
+              width="222"
+              alt="카카오 로그인 버튼"
+            />
+          </li>
+          <li onclick="kakaoLogout();">
+            <a href="javascript:void(0)">
+              <!-- <span>카카오 로그아웃</span> -->
+              <button style="background-color: white">카카오 로그아웃</button>
+            </a>
+          </li>
+        </ul>
```
ul 과 li 태그를 이런 상황에서 쓰는게 아닌 것 같은데요.. 각 태그들의 의미를 공부해보면 좋을 것 같습니다.

Enoch-Kim commented on apps/web-login/index.html:
```diff
@@ -0,0 +1,57 @@
+            </a>
+          </li>
+        </ul>
+        <!-- 카카오 스크립트 -->
+        <script src="https://developers.kakao.com/sdk/js/kakao.js"></script>
```
script 들의 위치를 정리해주면 좋을 것 같습니다. 그리고 script 들이 로드되는 방식에 대해 더 알아보면 좋을 것 같아요!

Enoch-Kim commented on apps/web-login/index.html:
```diff
@@ -0,0 +1,57 @@
+              alt="카카오 로그인 버튼"
+            />
+          </li>
+          <li onclick="kakaoLogout();">
+            <a href="javascript:void(0)">
+              <!-- <span>카카오 로그아웃</span> -->
+              <button style="background-color: white">카카오 로그아웃</button>
+            </a>
+          </li>
```
로그인과 로그아웃 버튼이 왜 한 UI에 같이 있는 것일까요? 로그인을 한 후에는 새로운 페이지로 넘어가지 않나요?

Enoch-Kim commented on apps/web-login/index.html:
```diff
@@ -0,0 +1,57 @@
+    <div class="login-box">
+      <h2>Login</h2>
+      <form>
+        <div class="user-box">
+          <input type="text" name="" required="" />
+          <label>Username</label>
+        </div>
+        <div class="user-box">
+          <input type="password" name="" required="" />
+          <label>Password</label>
+        </div>
+        <a href="#"> Submit </a>
```
이부분은 실제 사용하지 않는데 왜 있는 것일까요?

hihibara replied:
다음에 Oauth 말고 따로 회원 가입해서 user 정보를 저장하는 것을 구현해 보고 싶어서 일단 만들어 두었습니다

Enoch-Kim commented on apps/web-login/README.md:
```diff
@@ -0,0 +1,3 @@
+# web 로그인 구현 테스트
+
+1. 카카오 로그인 구현
```
카카오 로그인만 구현되어있나요? README를 더 자세히 써주시면 좋을 것 같아요

Enoch-Kim commented on apps/web-login/js/CallBack.js:
```diff
@@ -0,0 +1,19 @@
+const naverOuath_ID = "Hw7RwBFOB7UjSufSEEjy";
```
`constants.js` 파일을 두고 이런 정보들을 모아두면 좋을 것 같네요!

Enoch-Kim commented on apps/web-login/js/CallBack.js:
```diff
@@ -0,0 +1,19 @@
+const naverOuath_ID = "Hw7RwBFOB7UjSufSEEjy";
+const callBack_URL = new URL(document.location.href);
```
변수명의 규칙이 이상한 것 같습니다. 카멜케이스와 스네일 케이스가 왜 혼합되어있을까요?
```suggestion
const callBackUrl = new URL(document.location.href);
```

Enoch-Kim replied:
추가로 파일명도 명확하지 않은 것 같습니다. 왜 파일명을 CallBack.js로 두셨을까요?

hihibara replied:
#1 facebook이랑 google때 네이버 처럼 따로 callback.html이 있을 거 같아서 일부러 두었습니다

hihibara replied:
facebook이랑 google  Oauth 로그인을 만들 때도 naver처럼 따로 callBcak.html이 필요할 거 같아서 따로 만들었습니다

Enoch-Kim commented on apps/web-login/js/CallBack.js:
```diff
@@ -0,0 +1,19 @@
+const naverOuath_ID = "Hw7RwBFOB7UjSufSEEjy";
+const callBack_URL = new URL(document.location.href);
+const index_URL = callBack_URL.origin;
+var naver_id_login = new naver_id_login(naverOuath_ID, index_URL);
```
변수는 스네일 케이스로 할지 카멜 케이스로 할지 통일해주시면 좋을 것 같아요

hihibara replied:
혹시 #Enoch-Kim님은 어떤 방법으로 사용하시나요? 그냥 통일성 있게 하시나요?

hihibara replied:
@Enoch-Kim님은 어떠한 규칙을 사용하시나요? 그냥 통일성 있게 하시나요?

Enoch-Kim replied:
js 와 java는 카멜케이스 씁니다.

Enoch-Kim commented on apps/web-login/js/CallBack.js:
```diff
@@ -0,0 +1,19 @@
+// 네이버 사용자 프로필 조회
+naver_id_login.get_naver_userprofile("naverSignInCallback()");
+// 네이버 사용자 프로필 조회 이후 프로필 정보를 처리할 callback function
+function naverSignInCallback() {
+  alert(naver_id_login.getProfileData("email"));
+  alert(naver_id_login.getProfileData("nickname"));
+  alert(naver_id_login.getProfileData("age"));
+  alert("로그인 성공");
+}
```
로그인이 성공하면 다른 성공한 페이지로 넘어가주면 좋을 것 같습니다.

hihibara replied:
@Enoch-Kim  그러면 아예 홈 페이지 예제를 만들어서 해볼게요

hihibara replied:
@Enoch-Kim  다음 페이지로 넘어가는 것을 구현하려고 하는데 user 정보를 homepage 에서 볼 수 있게 하고 싶어요 
근데 이게 방향을 어떻게 가야 할지 모르겠습니다. 

서버에서 받은 정보를 어떻게 저장하는 것이 가장 좋을까요?

Enoch-Kim commented on apps/web-login/js/Login.js:
```diff
@@ -0,0 +1,70 @@
+// naver ouath part
+const WEB_IP = new URL(document.location.href);
```
어떤 곳은 WEB_IP 이고 어떤 곳은 callBack_URL 이네요?

Enoch-Kim commented on apps/web-login/js/Login.js:
```diff
@@ -0,0 +1,70 @@
+
+//카카오로그아웃
+function kakaoLogout() {
+  if (Kakao.Auth.getAccessToken()) {
+    Kakao.API.request({
+      url: "/v1/user/unlink",
+      success: function (response) {
+        console.log(response);
+      },
+      fail: function (error) {
+        console.log(error);
+      },
+    });
+    Kakao.Auth.setAccessToken(undefined);
+  }
```
토큰이 없을때 어떻게 행동할지 기술되어있지 않네요

Enoch-Kim replied:
왜 요청을 한 후에 undefined로 토큰을 설정하는 것일까요?

hihibara replied:
@Enoch-Kim 예제에 나와있는 것을 그대로 사용했는데요 이 부분 수정하겠습니다.

Enoch-Kim reviewed (commented):
분발합시다.

hihibara committed:
homepage추가, UI 변경

Enoch-Kim commented:
@hihibara 한 commit에 너무 많은 내용이 담긴 것 같아요. commit을 쪼개서 올려줄 수 있을까요?
예를들어 [ bootstrap 디펜던시 추가 / 홈 화면 추가 / 홈화면으로 렌더링하는 기능 구현 ] 이런 식으로 기능을 더 많이 쪼개서 commit을 분리해주면 좋을 것 같아요!

hihibara committed:
Naver user 정보를 cookies에 저장

hihibara committed:
로그인하면 index.html에서 user 닉네임 보여주기기

hihibara committed:
네이버 Logout 구현

hihibara committed:
naver token 삭제

hihibara committed:
블러그 Kakao Oauth

Enoch-Kim commented:
1. git reset soft
2. git commit 쪼개기
3. commit을 쪼개면서 기능별로 PR 나누기

hihibara committed:
버튼을 누를때만 로그인 되도록 구현

hihibara committed:
kakao 로그아웃 수정

hihibara committed:
constants.js & url_list.js생성해서 코드 리팩토링

hihibara committed:
Logout.js 생성후 naver & kakao 로그아웃 기능 코드 분할

hihibara committed:
<refactory> js/안에 module 파일 추가, 파일들 이름 변경, Index-Page.js 생성

[변경사항]
            -> js/Index-Page.js
js/Login.js -> js/Login-Page.js
js/CallBack.js -> js/NaverCallBack-Page.js
js/Logout.js -> js/moudle/Logout.js
js/constants.js -> js/moudle/constants.js
js/url_list.js -> js/moudle/url_list.js

refactory를 하면서 .js 파일을 추가로 더 생성을 하게 되었고 그 결과 많은 파일들로 인해서 파일을 찾는 효율이 줄어 들었고 파일 이름이
원래 부터 직관적이지 못해서 rename하게 되었습니다

hihibara committed:
<test> NaverCallback-Page.js애서 userprofile 정보를 못가져왔다

refactory를 하는 중에 userprfile에 원래있어야 할 정보가 undefine
으로 바뀐것을 확인했다. 그래서 원인 찾던중 NaverCallBack-Page.js가
module로 바뀐 다음이라고 생각이 들었고 moudule을 취소해보니 원래대로
작동하게 되었다.

hihibara committed:
<docs> user profile을 오류해결 정상작동

-secction storage를 활용하여서 기존에는 constants.js에 import 하여
받아오는 정보들을  NaverCallBack-Page.js를 module화 하지않고
사용하였습니다.

hihibara committed:
<docs> naver logout을 index.html에서 실행

- constants.js에 secctionStorage.setItem()을 사용 기존에 Login-Page.js에
  있던 secctionStorage.setItem() 코드 삭제
- Logout.js에 있는 naverLogOut()을 index-Page.html로 이동 (앞으로 계속
   index-Page.html에서 사용할 예정)
- logout기능에 쿠기와 토근 삭제 기능 있음음

Enoch-Kim commented on apps/web-login/login.html:
```diff
@@ -0,0 +1,69 @@
+        <div class="LogOutBtn" id="kakao-logout-btn">
+          <a href="javascript:void(0)">
+            <!-- <span>카카오 로그아웃</span> -->
+            <button style="background-color: white">카카오 로그아웃</button>
```
@hihibara  로그아웃 버튼이 이 페이지에 존재하는게 맞는건지 모르겠네요
![스크린샷 2023-04-11 오전 1 52 44](https://user-images.githubusercontent.com/60127664/230950617-61eb8099-0218-412e-9099-406a225c4c4f.png)

Enoch-Kim commented on apps/web-login/login.html:
```diff
@@ -0,0 +1,69 @@
+  <body>
+    <div class="login-box">
+      <h2>Login</h2>
+      <form>
+        <div class="user-box">
+          <input type="text" name="" required="" />
+          <label>Username</label>
+        </div>
+        <div class="user-box">
+          <input type="password" name="" required="" />
+          <label>Password</label>
+        </div>
+        <a id="submit" href="#"> Submit </a>
+      </form>
```
테스트용 아이디와 비번을 설정해두고 다음 페이지로 접속할 수 있게 해주시면 좋을 것 같아요

Enoch-Kim commented on apps/web-login/js/moudle/url_list.js:
```diff
@@ -0,0 +1,9 @@
+let indexURL = new URL(document.location.href);
+let pathname = indexURL.pathname; // url-path 부분
+
+let callbackURL =
+  indexURL.origin + // scheme와 hosts 부분
+  pathname.substring(0, pathname.lastIndexOf("/")) +
+  "/CallBack.html";
+
+export { indexURL, callbackURL, pathname };
```
이 부분은 상수가 아닌데 왜 따로 빼두었나요? 호출되는 부분을 확실히 알지 못하면 사이드이펙트가 발생할 가능성이 많습니다. 아니면 다음과 같이 함수로 빼둬서 호출 시점에 생성되도록 하는게 좋을 것 같네요.
```suggestion
function getIndexUrl() {
    return new URL(document.location.href)
}

function getCallbackUrl(indexUrl) {
    const origin = indexUrl.origin
    const pathname = indexUrl.pathname
    return origin + pathname.substring(0, pathname.lastIndexOf("/")) + "/CallBack.html"
}

export { getIndexUrl, getCallbackUrl };
```

Enoch-Kim commented on apps/web-login/js/moudle/constants.js:
```diff
@@ -0,0 +1,9 @@
+const NAVER_CLIENT_ID = "Hw7RwBFOB7UjSufSEEjy"; // 상수는 따로 .js 파일에 넣어서 관리하자
+const KAKAO_CLIENT_ID = "13cee8f0d511dff11f766fe1d031b0a1";
+const cookieName = "userProfile";
```
cookieName 만 camel-case로 설정한 이유가 있을까요?

Enoch-Kim commented on apps/web-login/js/moudle/constants.js:
```diff
@@ -0,0 +1,9 @@
+const NAVER_CLIENT_ID = "Hw7RwBFOB7UjSufSEEjy"; // 상수는 따로 .js 파일에 넣어서 관리하자
+const KAKAO_CLIENT_ID = "13cee8f0d511dff11f766fe1d031b0a1";
```
위의 데이터들은 보안상 위험한 키값인 것 같은데 어떻게 대처해야할까요?

Enoch-Kim commented on apps/web-login/js/moudle/constants.js:
```diff
@@ -0,0 +1,9 @@
+const KAKAO_CLIENT_ID = "13cee8f0d511dff11f766fe1d031b0a1";
+const cookieName = "userProfile";
+const NAVER_OAUTH_SECRETE = "MaLARH3T4B";
+sessionStorage.setItem("cookieName", `${cookieName}`);
+sessionStorage.setItem("NAVER_CLIENT_ID", `${NAVER_CLIENT_ID}`);
+sessionStorage.setItem("NAVER_OAUTH_SECRETE", `${NAVER_OAUTH_SECRETE}`);
+
+export { NAVER_CLIENT_ID, KAKAO_CLIENT_ID, cookieName, NAVER_OAUTH_SECRETE };
```
이미 export 했는데 왜 상수를 세션 스토리지에 넣어서 사용하는걸까요?

Enoch-Kim replied:
그리고 string 을 넣는데 왜 다시 ``로 덮어서 사용하는거죠?
```suggestion
sessionStorage.setItem("cookieName", cookieName);
sessionStorage.setItem("NAVER_CLIENT_ID", NAVER_CLIENT_ID);
sessionStorage.setItem("NAVER_OAUTH_SECRETE", NAVER_OAUTH_SECRETE);

export { NAVER_CLIENT_ID, KAKAO_CLIENT_ID, cookieName, NAVER_OAUTH_SECRETE };
```

Enoch-Kim commented on apps/web-login/js/moudle/Logout.js:
```diff
@@ -0,0 +1,50 @@
+        console.log(error);
+      },
+    });
+    Kakao.Auth.setAccessToken(undefined);
```
왜 undefined로 토큰 설정을 하는걸까요?

Enoch-Kim commented on apps/web-login/js/moudle/Logout.js:
```diff
@@ -0,0 +1,50 @@
+  }
+}
+
+function DelNaverToken(clentID, secrete, token) {
+  const delTokenURL =
+    `https://nid.naver.com/oauth2.0/token?grant_type=delete&client_` +
+    `id=${clentID}&client_` +
+    `secret=${secrete}&access_` +
+    `token=${token}&service_provider=NAVER`;
+  location.replace(delTokenURL);
+
+  // window.close();
+}
+
+function ChangeLoginBox() {
+  loginBox.innerHTML = `<a type="module" href="login.html">Log In</a>`;
+  location.reload();
+}
```
왜 이 함수들의 이름은 대문자로 시작하나요?

Enoch-Kim commented on apps/web-login/js/NaverCallBack-Page.js:
```diff
@@ -0,0 +1,62 @@
+  }
+}
+
+function UserProfileInputCookies() {
+  userProfile = {
+    token: naverIdLogin.oauthParams.access_token,
+    email: naverIdLogin.getProfileData("email"),
+    nickname: naverIdLogin.getProfileData("nickname"),
+    age: naverIdLogin.getProfileData("age"),
+  };
+
+  if (!userProfile) {
+    return false;
+  } else {
+    document.cookie = cookieName + "=" + JSON.stringify(userProfile);
+    console.log("userProfile", userProfile);
+    return true;
+  }
+}
+
+function CloseCallbackPage() {
+  window.opener.postMessage(
+    "use Naver Oauth",
+    indexURL + // scheme와 hosts 부분
+      pathname.substring(0, pathname.lastIndexOf("/")) +
+      "/login.html"
+  );
+  window.open("http://127.0.0.1:5500/level-up-test/apps/web-login/index.html");
+  window.close();
+}
+
+function LogOut() {
+  window.open(`https://nid.naver.com/nidlogin.logout`);
+}
```
대문자로 시작하는 함수가 따로 있는 이유가 있을까요?

Enoch-Kim reviewed (changes requested):
@hihibara 네이밍 컨벤션 신경써주세요
